### PR TITLE
[Perf][GLM-5.2] Optimize quantized MTP serving

### PR DESCRIPTION
**Pull request by WoosukKwon**
## Purpose

Optimize quantized GLM-5.2 serving on NVIDIA GPUs while preserving model
behavior and the existing dense MLP path.

This change:

- fuses activation and dynamic FP8 quantization, including the DeepGEMM
  SiLU-multiply path;
- fuses FlashInfer all-reduce + residual + RMSNorm + packed group-FP8
  quantization and forwards the prequantized activation into attention or MoE;
- lets supported modular MoE backends consume prequantized activations without
  repeating input quantization;
- keeps DeepSeek V3.2 / GLM-5.2 MTP on the compile-free NVIDIA implementation,
  preserves separate logits and feedback hidden-state contracts, and propagates
  quantized activations through speculative decoding;
- retains main's MRV2 fused multi-step decode and shared MTP top-k index buffer;
  and
- leaves dense MLP execution unchanged.

## Duplicate-work check

No issue number was provided, so an issue-specific PR search was not applicable.
I searched open PRs for `GLM-5.2 quant fusion MTP`,
`allreduce RMSNorm FP8 quant`, `DeepSeek V3.2 MTP optimization`, and
`dynamic FP8 fusion`.

The closest open work is adjacent rather than duplicative:

- #42230 adds a standalone SymmetricMemory Triton collective; this PR integrates
  FlashInfer's collective epilogue with GLM-5.2, DeepGEMM, MoE, and MTP.
- #45364 manually fuses Llama RMSNorm-to-quant without an all-reduce or
  DeepGEMM/MoE data flow.
- #45770 fixes mixed RMSNorm input/weight dtypes in existing kernels.
- #49828 adds a standalone CUDA SiLU-and-multiply dynamic per-token quant kernel;
  this PR's activation contract and DeepGEMM packed group-FP8 path are different.
- #45151 fuses an attention epilogue, and #46935 is an AsyncTP scaffold; neither
  implements this end-to-end model path.

## Performance and model evaluation

Hardware: two GB200 nodes, TP=8, concurrency=1, random 8,192-token input and
1,024-token output, two warmups, ten measured prompts, temperature 0, MTP=3,
piecewise CUDA graphs.

| zai-org/GLM-5.2-FP8 | Main | Optimized | Change |
|---|---:|---:|---:|
| Mean TTFT | 587.68 ms | 544.75 ms | -7.3% |
| Mean TPOT | 76.80 ms | 74.35 ms | -3.2% |
| MTP acceptance | 83.70% | 87.07% | +3.37 pp |

This A/B was collected before the final rebase onto current main; the resolved
rebase retained main's newer MRV2 lifecycle.

The final compile-free acceptance validation used the same 8K/1K, C1, MTP=3,
TP=8 setup and completed all ten prompts without failures:

| Model | Overall | Position 0 | Position 1 | Position 2 |
|---|---:|---:|---:|---:|
| zai-org/GLM-5.2-FP8 | 83.47% | 90.83% | 83.00% | 76.57% |
| nvidia/GLM-5.2-NVFP4 | 89.82% | 93.79% | 90.15% | 85.53% |

Full GSM8K evaluation for `zai-org/GLM-5.2-FP8`: 1,242 / 1,319 strict exact
matches, **94.16%**. The same score was reproduced after the miscellaneous quant
fusions and after the autoregressive/MTP changes.

## Test plan and results

```bash
.venv/bin/python -m pytest \
  tests/config/test_speculative_draft_hf_overrides.py \
  tests/fusion/test_quant_activation_contract.py \
  tests/models/deepseek_v32/test_sequence_parallel.py \
  tests/v1/spec_decode/test_llm_base_proposer_sampling.py -q
# 25 passed

.venv/bin/python -m pytest \
  tests/v1/worker/test_gpu_autoregressive_speculator.py -q
# 17 passed

.venv/bin/python -m pytest \
  tests/kernels/core/test_fused_allreduce_gemma_rms_norm.py::test_fused_allreduce_rms_norm_group_fp8 -q
# 4 passed

.venv/bin/python -m pytest \
  tests/kernels/moe/test_deepgemm.py \
  tests/kernels/moe/test_silu_mul_fp8_quant_deep_gemm.py \
  tests/kernels/test_fused_deepseek_v32_norm_rope.py -q
# 85 passed, 3 skipped

git diff --name-only -z origin/main...HEAD | xargs -0 pre-commit run --files
# all applicable hooks passed, including ruff and mypy
```

A broader run of `test_fused_allreduce_gemma_rms_norm.py` passed 18 existing
distributed cases, then hit one unchanged BF16 Gemma comparison at TP=4:
15 / 1,363,968 residual elements differed, with max absolute difference 0.03125
against the existing 0.02 tolerance. The newly added packed FP8 cases were then
run independently and all four passed. No tolerance was changed in this PR.

## AI assistance disclosure

This change was developed with OpenAI Codex assistance. This is intentionally a
draft PR. Before marking it ready for review, the human submitter must review
every changed line, understand and be able to defend the implementation and
results end-to-end, and update this section to confirm that review.

### CODE REVIEW

**Comment by WoosukKwon** [comment]
Superseded by focused draft PRs:

- #51938 — sparse DeepSeek V3.2 MTP registration (independent)
- #51939 — DeepGEMM block-linear prequantized inputs (independent)
- #51941 — modular MoE prequantized DeepGEMM inputs (independent)
- #51942 — FlashInfer all-reduce/RMSNorm/packed-FP8 primitive (independent)
- #51943 — DeepSeek V3.2 attention integration (depends on #51939 and #51942)
- #51945 — autoregressive model integration (depends on #51939, #51941, #51942, and #51943)
- #51947 — MTP LM-head integration (depends on #51938, #51939, and #51942)

All 42 files from this draft are accounted for. During the split I also dropped the misleading compile-free test-only assertion and the unrelated external-router rewrite. The replacement PRs retain the relevant validation and AI-assistance disclosures.
